src/execution_plan/execution_plan.c

```diff
@@ -875,42 +860,44 @@ ExecutionPlan *NewExecutionPlan(ResultSet *result_set) {
 	array_free(segment_indices);
 
 	ExecutionPlan *plan = segments[segment_count - 1];
-
 	// The root operation is OpResults only if the query culminates in a RETURN or CALL clause.
-	if(query_has_return) {
-		if(!connecting_op) {
-			// Set the connecting op if our query is just a RETURN.
-			assert(segment_count == 1);
-			connecting_op = ExecutionPlan_LocateFirstOp(plan->root, OPType_PROJECT | OPType_AGGREGATE);
-		}
+	if(query_has_return || last_clause_type == CYPHER_AST_CALL) {
+		OpBase *results_op = NewResultsOp(plan);
+		_ExecutionPlan_UpdateRoot(plan, results_op);
+	}
 
-		// Prepare column names for the ResultSet.
-		if(result_set) ResultSet_SetColumns(result_set, AST_BuildColumnNames(last_clause));
+	// Disregard self.
+	plan->segment_count = segment_count - 1;
+	plan->segments = segments;
 
-		OpBase *results_op = NewResultsOp(plan, result_set);
-		_ExecutionPlan_UpdateRoot(plan, results_op);
+	return plan;
+}
+
+static void _ExecutionPlan_SetResultColumns(ExecutionPlan *plan) {
```

Comment:
It seems like this logic is a bit miss placed,
To me it makes more sense to place setting of the result-set columns within `ResultSet_New`.